ksqldb-engine/src/main/java/io/confluent/ksql/schema/ksql/inference/SchemaRegisterInjector.java

```diff
@@ -155,80 +167,121 @@ private void registerSchemas(
+      final SchemaAndId schemaAndId,
       final String topic,
       final FormatInfo formatInfo,
-      final SerdeFeatures serdeFeatures,
       final KsqlConfig config,
       final String statementText,
-      final boolean registerIfSchemaExists,
       final String subject,
-      final boolean isKey
+      final Boolean isKey
   ) {
+    final String schemaIdPropStr = isKey ? "key_schema_id" : "value_schema_id";
     final Format format = FormatFactory.of(formatInfo);
-    if (!format.supportsFeature(SerdeFeature.SCHEMA_INFERENCE)) {
-      return;
+    if (!canRegister(format, config, topic)) {
+      throw new KsqlStatementException(schemaIdPropStr + " is provided but format "
+          + format.name() + " doesn't support registering in Schema Registry",
+          statementText);
     }
 
-    if (config.getString(KsqlConfig.SCHEMA_REGISTRY_URL_PROPERTY).isEmpty()) {
-      throw new KsqlSchemaRegistryNotConfiguredException(
-          String.format(
-              "Cannot create topic '%s' with format %s without configuring '%s'",
-              topic, format.name(), KsqlConfig.SCHEMA_REGISTRY_URL_PROPERTY)
-      );
+    final SchemaTranslator translator = format.getSchemaTranslator(formatInfo.getProperties());
+    if (!translator.name().equals(schemaAndId.rawSchema.schemaType())) {
+      throw new KsqlStatementException(String.format(
+          "Format and fetched schema type using %s %d are different. Format: [%s], "
+              + "Fetched schema type: [%s].",
+          schemaIdPropStr, schemaAndId.id, format.name(), schemaAndId.rawSchema.schemaType()),
+          statementText);
     }
 
-    try {
-      final SchemaRegistryClient srClient = serviceContext.getSchemaRegistryClient();
+    final int id = registerRawSchema(serviceContext.getSchemaRegistryClient(),
+        schemaAndId.rawSchema, topic, subject, statementText, isKey);
 
-      if (registerIfSchemaExists || !SchemaRegistryUtil.subjectExists(srClient, subject)) {
-        final SchemaTranslator translator = format.getSchemaTranslator(formatInfo.getProperties());
+    if (id != schemaAndId.id) {
+      throw new KsqlStatementException(
+          "Schema id registered is "
+              + id
+              + " which is different from provided " + schemaIdPropStr + " " + schemaAndId.id + "."
+              + System.lineSeparator()
+              + "Topic: " + topic
+              + System.lineSeparator()
+              + "Subject: " + subject
+              + System.lineSeparator()
+              + "Schema: " + schemaAndId.rawSchema, statementText
+      );
+    }
+  }
 
-        final ParsedSchema parsedSchema = translator.toParsedSchema(
-            PersistenceSchema.from(schema, serdeFeatures)
+  private static int registerRawSchema(
+      final SchemaRegistryClient srClient,
+      final ParsedSchema parsedSchema,
+      final String topic,
+      final String subject,
+      final String statementText,
+      final boolean isKey
+  ) {
```

Comment:
This method has the potential to be moved to `SchemaRegistryUtil`. I see different methods there that deal with authentication errors (like this method). What do you think? You can write unit tests for that method in `SchemaRegistryUtilTest`.

If you move it, the only parameter I think we should not move is the `statementText`, which is only used to customize the error message. We should catch the exception when calling the method instead, and re-throw it with the statement text in this class.

Reply:
Yeah. Let me take a look and put it there